app/src/rgb_underglow.c

```diff
@@ -0,0 +1,295 @@
+};
+
+struct led_hsb {
+	u16_t h;
+	u8_t  s;
+	u8_t  b;
+};
+
+struct rgb_underglow_state {
+    u16_t hue;
+    u8_t  saturation;
+    u8_t  brightness;
+    u8_t  animation_speed;
+    u8_t  current_effect;
+    u16_t animation_step;
+    bool  on;
+};
+
+struct rgb_underglow_state state;
+
+struct device *led_strip;
+
+struct led_rgb pixels[STRIP_NUM_PIXELS];
+
+static struct led_rgb hsb_to_rgb(struct led_hsb hsb)
+{
+    double r, g, b;
+
+    u8_t i = hsb.h / 60;
+    double v = hsb.b / 100.0;
+    double s = hsb.s / 100.0;
+    double f = hsb.h / 360.0 * 6 - i;
+    double p = v * (1 - s);
+    double q = v * (1 - f * s);
+    double t = v * (1 - (1 - f) * s);
+
+    switch (i % 6)
+    {
+        case 0: r = v; g = t; b = p; break;
+        case 1: r = q; g = v; b = p; break;
+        case 2: r = p; g = v; b = t; break;
+        case 3: r = p; g = q; b = v; break;
+        case 4: r = t; g = p; b = v; break;
+        case 5: r = v; g = p; b = q; break;
+    }
+
+    struct led_rgb rgb = { r: r*255, g: g*255, b: b*255 };
+
+    return rgb;
+}
+
+static void zmk_rgb_underglow_effect_solid()
+{
+    for (int i=0; i<STRIP_NUM_PIXELS; i++)
+    {
+        int hue = state.hue;
+        int sat = state.saturation;
+        int brt = state.brightness;
+
+        struct led_hsb hsb = { hue, sat, brt };
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+}
+
+static void zmk_rgb_underglow_effect_breathe()
+{
+    for (int i=0; i<STRIP_NUM_PIXELS; i++)
+    {
+        int hue = state.hue;
+        int sat = state.saturation;
+        int brt = abs(state.animation_step - 1200) / 12;
+
+        struct led_hsb hsb = { hue, sat, brt };
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+
+    state.animation_step += state.animation_speed * 10;
+    
+    if (state.animation_step > 2400) {
+        state.animation_step = 0;
+    }
+}
+
+static void zmk_rgb_underglow_effect_spectrum()
+{
+    for (int i=0; i<STRIP_NUM_PIXELS; i++)
+    {
+        int hue = state.animation_step;
+        int sat = state.saturation;
+        int brt = state.brightness;
+
+        struct led_hsb hsb = { hue, sat, brt };
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+
+    state.animation_step += state.animation_speed;
+    state.animation_step = state.animation_step % 360;
+}
+
+static void zmk_rgb_underglow_effect_swirl()
+{
+    for (int i=0; i<STRIP_NUM_PIXELS; i++)
+    {
+        int hue = (360 / STRIP_NUM_PIXELS * i + state.animation_step) % 360;
+        int sat = state.saturation;
+        int brt = state.brightness;
+
+        struct led_hsb hsb = { hue, sat, brt };
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+
+    state.animation_step += state.animation_speed * 2;
+    state.animation_step = state.animation_step % 360;
+}
+
+static void zmk_rgb_underglow_tick(struct k_work *work)
+{
+    switch (state.current_effect)
+    {
+        case UNDERGLOW_EFFECT_SOLID:
+            zmk_rgb_underglow_effect_solid();
```

Comment:
Perhaps future thought... but for size sake later, we should probably make each effect be enabled/disabled via Kconfig setting.